### PR TITLE
🔢 Feat / Stats Calculator POC

### PR DESCRIPTION
**Pull request by demariadaniel**
- Adds basic Node module which queries Iobio server and outputs JSON
- Adds utilities for obtaining statistics data, histogram data, and calculating mean Read Coverage
- Better separates Client and Node TSconfigs
- URL entered via command line. demo data url: https://s3.amazonaws.com/iobio/NA12878/NA12878.autsome.bam

https://app.zenhub.com/workspaces/overture-stack-5d2e058ff67cc800011fee6b/issues/gh/overture-stack/iobio-components/31

### CODE REVIEW

File: packages/iobio-react-components/NA12878.autsome.bam statistics.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"statistics":{"mapped_reads":89988,"mapped_reads_percentage":0.9972,"forward_strands":45254,"forward_strands_percentage":0.5015,"proper_pairs":88192,"proper_pairs_percentage":0.9773,"singletons":265,"singletons_percentage":0.002937,"both_mates_mapped":89723,"both_mates_mapped_percentage":0.9942,"duplicates":422,"duplicates_percentage":0.004676,"failed_qc":0,"first_mates":44990,"last_read_position":140394619,"paired_end_reads":90242,"reverse_strands":44988,"second_mates":45252,"total_reads":90242}}
```
Comment (demariadaniel) [inline packages/iobio-react-components/NA12878.autsome.bam statistics.json:1]: Example file

File: packages/iobio-react-components/src/stats_util.mts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/*
```
Comment (demariadaniel) [inline packages/iobio-react-components/src/stats_util.mts:1]: Main Stats utility

follow up PR to start adding folder organization separating FE / BE concerns:
https://github.com/overture-stack/iobio-components/pull/45

File: packages/iobio-react-components/NA12878.autsome.bam statistics.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (demariadaniel) [inline packages/iobio-react-components/NA12878.autsome.bam statistics.json:1]: Demo file. Doesn't need to be committed but included for review.

File: packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+		"failed_qc": 0,
+		"first_mates": 49438,
+		"last_read_position": 52662640,
+		"paired_end_reads": 98905,
+		"reverse_strands": 49312,
+		"second_mates": 49467,
+		"total_reads": 98905,
+		"mean_read_coverage": 52
+	}
+}
```
Comment (demariadaniel) [inline packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: Demo file included for review. I can add `statistics-*.json` type files to gitignore if this looks good

File: packages/iobio-react-components/src/stats_util.mts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ *  TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ *  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ *  IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+import fs from 'fs';
+
+// Requires Node 19
```
Comment (ciaranschutte) [inline packages/iobio-react-components/src/stats_util.mts:24]: worth setting this as required Node version 
https://stackoverflow.com/questions/29349684/how-can-i-specify-the-required-node-js-version-in-package-json

File: packages/iobio-react-components/src/stats_util.mts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+const db = new DataBroker(fileUrl);
+
+const data: any[] = [];
+
+db.addEventListener('stats-stream-start', () => {
+	console.log('Streaming started');
+});
+
+db.addEventListener('stats-stream-data', (evt: any) => {
```
Comment (ciaranschutte) [inline packages/iobio-react-components/src/stats_util.mts:42]: `evt` can be `event`
I forget how @justincorrigible elegantly phrases it - but essentially use full words for variables names (for the most part)

File: packages/iobio-react-components/src/utils.ts [hunk 7]
```diff
@@ -52,3 +62,68 @@ export const setElementStyles = (element: Element, styles: string) => {
...
+			if (percentKeys.some((percentKey) => percentKey === val)) {
+				const percentage = Number((value / dataEvent['total_reads']).toPrecision(4));
+				const key = `${val}_percentage`;
+				stats[key] = percentage;
+			}
+			return stats;
+		},
+		{} as { [k: string]: number },
+	);
+};
```
Comment (ciaranschutte) [inline packages/iobio-react-components/src/utils.ts:91]: would be good to explicitly name `(acc, val)` to what they are
for this looks like it's something like acc => stats and val => dataKey ?

File: packages/iobio-react-components/src/stats_util.mts (hunk 5 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/stats_util.mts:24]: Thanks I'll give that a look

File: packages/iobio-react-components/src/utils.ts (hunk 7 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/utils.ts:91]: Yes explicit names are always preferred, will revise

File: packages/iobio-react-components/src/stats_util.mts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+import fs from 'fs';
+
+// Requires Node 19
+import { DataBroker } from 'iobio-charts/data_broker.js';
+
+import { calculateMeanCoverage, getBamStatistics } from './utils.ts';
+
+const fileUrl = 'https://s3.amazonaws.com/iobio/NA12878/NA12878.autsome.bam';
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply packages/iobio-react-components/src/stats_util.mts]: Updated to accept URL in command line argument `pnpm run stats ${fileUrl}`

File: packages/iobio-react-components/src/utils.ts (hunk 7 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/utils.ts]: thanks! updated!

File: packages/iobio-react-components/src/stats_util.mts (hunk 6 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/stats_util.mts]: updated!

File: packages/iobio-react-components/src/stats_util.mts (hunk 5 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/stats_util.mts]: OK added engine to package.json + npmrc
Updated to Node 22, which is LTS https://nodejs.org/en/about/previous-releases#looking-for-latest-release-of-a-version-branch 
Required some syntax changes but working now. Good improvements in Node releases recently.

File: packages/iobio-react-components/package.json [hunk 9]
```diff
@@ -10,7 +10,7 @@
 		"build": "tsc -b && vite build",
 		"lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
 		"preview": "vite preview",
-		"stats": "pnpm exec ts-node -P tsconfig.node.json --esm src/stats_util.mts"
+		"stats": "pnpm exec tsx src/stats_util.mts"
```
Comment (ciaranschutte) [inline packages/iobio-react-components/package.json:13]: can likely remove `ts-node` dependency if not using it any more?

File: packages/iobio-react-components/src/stats_util.mts [hunk 10]
```diff
@@ -44,7 +43,7 @@ db.addEventListener('stats-stream-data', (event: any) => {
 	data.push(event.detail);
 });
 
-db.addEventListener('stats-stream-end', () => {
+db.addEventListener('stats-stream-end', async () => {
```
Comment (ciaranschutte) [inline packages/iobio-react-components/src/stats_util.mts:46]: What has become async here?

Comment (demariadaniel) [reply packages/iobio-react-components/src/stats_util.mts]: Artifact left from debugging / updating `fs` syntax, removed

File: packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json (hunk 4 above)
Comment (demariadaniel) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: Added to gitignore

Comment (ciaranschutte) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: unclear, do you want this merged or ignored?

Comment (demariadaniel) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: Sorry for confusion. This is not intended to be committed, but I did want to see if there were any opinions or changes requested on the output.
@justincorrigible any thoughts?

File: packages/iobio-react-components/src/utils.ts [hunk 11]
```diff
@@ -52,3 +62,68 @@ export const setElementStyles = (element: Element, styles: string) => {
...
+/**
+ * Obtain BAM Histogram data from Data Broker data events
+ * @param dataEvent { [K in BamHistogramKey]: { [numKey: string]: number } }
+ * where numKey parses to an Integer
+ */
+
+type HistogramData = { [K in BamHistogramKey]: { [numKey: string]: number } };
+
+export const getHistogramData = (dataEvent: HistogramData) => {
+	const histogramData = histogramKeys.reduce((acc, val) => {
```
Comment (justincorrigible) [inline packages/iobio-react-components/src/utils.ts:102]: missed the `acc, val` declarative renaming here

File: packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json (hunk 4 above)
Comment (justincorrigible) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: LGTM! I'd have preferred camelCase keys, as per JS convention, but I'd imagine these are just reflecting iobio internals. correct?

File: packages/iobio-react-components/README.md
Comment (justincorrigible) [inline packages/iobio-react-components/README.md:1]: worth adding a bit later about how this is a monorepo (for those more who cannot see it obviously), `pnpm` being required, how to install it, etc.

File: packages/iobio-react-components/.gitignore [hunk 12]
```diff
@@ -21,3 +21,5 @@ dist-ssr
 *.njsproj
 *.sln
 *.sw?
+
+statistics-*.json
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply packages/iobio-react-components/.gitignore]: Empty lines should be added back but not seeing them displayed properly on Github 
https://github.com/overture-stack/iobio-components/pull/40

File: packages/iobio-react-components/package.json [hunk 13]
```diff
@@ -25,8 +27,12 @@
 		"eslint": "^8.57.0",
 		"eslint-plugin-react-hooks": "^4.6.2",
 		"eslint-plugin-react-refresh": "^0.4.7",
+		"tsx": "^4.19.1",
 		"typescript": "^5.5.2",
 		"vite": "^5.3.1",
 		"vite-plugin-dts": "^3.9.1"
+	},
+	"engines": {
+		"node": ">=22"
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply packages/iobio-react-components/package.json:36]: I've added the Todo comment with notes about Node 22. 
Node 20+ is required for this feature and engines was added as part of code review.

File: packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json (hunk 4 above)
Comment (demariadaniel) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json:24]: Yes these are straight from the iobio response

Comment (demariadaniel) [reply packages/iobio-react-components/statistics-NA12878.autsome.bam-2024-09-16.json]: Demo file removed

File: packages/iobio-react-components/src/utils.ts (hunk 11 above)
Comment (demariadaniel) [reply packages/iobio-react-components/src/utils.ts]: Updated

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**